geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/ReplayKafkaConsumerFeatureSource.scala

```diff
@@ -0,0 +1,419 @@
+      .map(ts => (ts, indexAtTime(ts)))
+      .getOrElse((replayConfig.end.getMillis, Some(0)))
+
+    startIndex.map { si =>
+      val endTime = startTime - replayConfig.readBehind.getMillis
+      snapshot(startTime, si, endTime)
+    }.getOrElse(None)
+  }
+
+  /** @return the index of the most recent [[GeoMessage]] at or before the given ``time``
+    */
+  private[kafka] def indexAtTime(time: Long): Option[Int] = {
+
+    if (replayConfig.isInWindow(time)) {
+      // it doesn't matter what the message is, only the time
+      val key: GeoMessage = new Clear(new Instant(time))
+
+      // reverse ordering for reverse ordered ``messages``
+      val ordering = new Ordering[GeoMessage] {
+        override def compare(x: GeoMessage, y: GeoMessage): Int = y.timestamp.compareTo(x.timestamp)
+      }
+
+      var index = java.util.Arrays.binarySearch(messages, key, ordering)
+
+      if (index < 0) {
+        // no message found at sought time
+        index = -index - 1
+      }
+
+      // there may be multiple messages at the same time so there may be messages before ``index``
+      // that are at the given ``time``
+      while (index > 0 && messages(index - 1).timestamp.getMillis <= time) index -= 1
+
+      if (index < messages.length) Some(index) else None
+    } else {
+      // requested time is outside of user specified time window
+      None
+    }
+  }
+
+  /**
+    * @param startIndex the index of the most recent message to process
+    * @param endTime the time of the last message to process
+    */
+  private def snapshot(time: Long, startIndex: Int, endTime: Long): Option[ReplaySnapshotFeatureCache] = {
+    val snapshot: Seq[GeoMessage] = messages.view
+      .drop(startIndex)
+      .takeWhile {
+        // stop at the first clear or when past the endTime
+        case c: Clear => false
+        case e => e.timestamp.getMillis >= endTime
+      }
+
+    if (snapshot.isEmpty) {
+      None
+    } else {
+      Some(ReplaySnapshotFeatureCache(replaySFT, time, snapshot))
+    }
+  }
+
+  private def readMessages(): Array[GeoMessage] = {
+
+    logger.debug("Begin reading messages from {} using {}", topic, replayConfig)
+    val readStart = System.currentTimeMillis()
+
+    // don't want to block waiting for more messages so specify  a timeout
+    val kafkaConsumer = kf.kafkaConsumer(topic, Map("consumer.timeout.ms" -> "250"))
+
+    // use the liveSFT to decode becuase that's the type that was used to encode
+    val msgDecoder = new KafkaGeoMessageDecoder(liveSFT)
+
+    // start 1 ms earlier because there might be multiple messages with the same timestamp
+    val startTime = replayConfig.realStartTime.minus(1L)
+
+    val offsetRequest = FindOffset(msg => {
+      val key = msgDecoder.decodeKey(msg)
+      if (key.ts.isEqual(startTime)) 0 else if (key.ts.isAfter(startTime)) 1 else -1
+    })
+
+    // required: there is only 1 partition;  validate??
+    val stream = kafkaConsumer.createMessageStreams(1, offsetRequest).head
```

Comment:
That what I thought but there is a configuration parameters for the number of partitions.  Why was that parameter added?  Could we support muliple partitions if messages are partitioned by the simple feature ID?  If so, replay would need some more work to handle.  Should the number of partitions parameter be removed or should I create a ticket for reviewing this another time?